src/components/my/Menu.tsx

```diff
@@ -8,16 +8,24 @@ import InternalLink from '../common/InternalLink';
 export interface MenuProps {
   label: string;
   rightElement?: ReactElement;
+  /**
+   * Internal Route에 사용됩니다.
+   * ex) /my/account, '/add/tag'
+   */
   href?: RouterPathType;
+  /**
+   * 새창열기에서 사용됩니다.
+   * like) Window.open target="_blank"
+   */
+  url?: string;
   onClick?: VoidFunction;
 }
```

Comment:
현재 받아야하는 props의 형태가
1. url, href, onClick 셋 다 없거나
2. 셋 중 하나만 있는 상태
로 파악이 되는 데 맞을까요 ??

맞다면 이를 더 type safe하게 나타낼 수 있을 거 같아요 !!

```suggestion
interface MenuBaseProps {
  label: string;
  rightElement?: ReactElement;
}

interface MenuInternalProps extends MenuBaseProps {
  href: RouterPathType;
}

interface MenuExternalProps extends MenuBaseProps {
  url: string;
}

interface MenuOnClickProps extends MenuBaseProps {
  onClick: VoidFunction;
}

type MenuProps = MenuBaseProps | MenuInternalProps | MenuExternalProps | MenuOnClickProps;
```

이렇게 하면 셋 중 하나만 들어가던가, 셋 다 없거나로 추론될 거 같아요!! (물론 아직까지 예상임니다 😢 )

Reply:
와 세상에....! ts 요렇게 사용하면 진짜 요기나겠네요! 
이거 진짜 오늘 머리 탁치고 갑니다!

Reply:
리뷰 적용완료 했습니다 3ef7ea6

만약 혜성님이 의도하신것 처럼 사용하려면 사용간에 사용하는 프롭은 never처리를 해주어야되더라구요!

```ts
interface MenuInternalProps extends MenuBaseProps {
  internalHref: RouterPathType;
  externalHref: never;
  onClick: never;
}
```

좀 더 알아보니 공공연하기 XOR이라는 Type을 만들어서 사용한다고 합니다. 그래서 코드는 아래와 같이 됩니다.
사용하지 않은 이유는 지금 당장 다른 곳에 이것이 쓰일곳이 없으며, 코드의 이해하려면 시간이 걸리기 떄문입니다.
따라서 이렇게 이슈화 시켜논다음 모든 팀원이 이해 이후 적용되면 좋을 것같다는 생각을 했습니다. 
왜냐하면 다른곳에서도 쓰시는 일이 분명히 생길 것같아서 입니다!! 

```ts
interface MenuBaseProps {
  label: string;
  rightElement?: ReactElement;
}
interface MenuInternalProps extends MenuBaseProps {
  internalHref: RouterPathType;
}

interface MenuExternalProps extends MenuBaseProps {
  externalHref: string;
}
interface MenuOnClickProps extends MenuBaseProps {
  onClick: VoidFunction;
}

export type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };
export type XOR<T, U> = T | U extends object ? (Without<T, U> & U) | (Without<U, T> & T) : T | U;

type MenuInternalPropsType = MenuBaseProps &
  XOR<MenuInternalProps, MenuExternalProps> &
  XOR<MenuInternalProps, MenuOnClickProps>;

type MenuExternalPropsType = MenuBaseProps &
  XOR<MenuExternalProps, MenuInternalProps> &
  XOR<MenuExternalProps, MenuOnClickProps>;

type MenuOnClickPropsType = MenuBaseProps &
  XOR<MenuOnClickProps, MenuInternalProps> &
  XOR<MenuOnClickProps, MenuExternalProps>;

type MenuProps =
  | MenuBaseProps
  | MenuInternalPropsType
  | MenuExternalPropsType
  | MenuOnClickPropsType;

const a: MenuInternalPropsType = { // success
  label: '',
  internalHref: '/',
};

const a: MenuInternalPropsType = { // error not `internalHref`
  label: '',
};
```


### 내부 사용에서는 이런식으로 나름의 타입 추론? 할 수 있도록해서 사용해서 했습니다.
```tsx
export default function Menu(props: MenuProps) {
 ...
 const { internalHref } = props as MenuInternalProps;
 ...
}
```

Reply:
https://timhwang21.gitbook.io/index/programming/typescript/xor-type
https://stackoverflow.com/questions/58630750/react-typescript-component-with-two-different-prop-interfaces